assets/aml-benchmark/components/src/dataset_preprocessor/dataset_preprocessor.py

```diff
@@ -0,0 +1,136 @@
+            mssg = (
+                "Please provide the input to apply preprocessing logic either via template input or \
+                script_path."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+        if self.user_preprocessor and not self.user_preprocessor.endswith('.py'):
+            mssg = (
+                "Please provide python script containing your custom preprocessor logic."
+            )
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg)
+            )
+
+    def add_json_filter(self, template):
+        """Add tojson filter in the template."""
+        patt = r'({{.*?}})'
+        pat = re.compile(patt)
+        matches = re.findall(pat, template)
+        for m in range(len(matches)):
+            logger.info(f"{template}, {matches[m]}")
+            s1 = '{{'+matches[m].lstrip('{{').rstrip('}}')+'|tojson}}'
+            template = template.replace(matches[m], s1)
+        logger.info(f"Final template:{template}")
+        return template
+
+    def prep_using_template(self):
+        """Preprocessor run using template."""
+        data = read_jsonl_files([self.input_dataset])
+        template = json.dumps(self.template)
+        template = json.loads(template)
+        template = self.add_json_filter(template)
+        env = jinja2.Environment()
+        jinja_template = env.from_string(template)
+        with open(self.output_dataset, mode='w', encoding='utf8') as f:
+            for example in data:
+                out = jinja_template.render(example)
+                out = out.replace('\'', '\"')
+                out_dict = json.loads(out)
+                if self.encoder_config:
+                    col_to_encode = self.encoder_config.get('column_name')
+                    out_dict[col_to_encode] = self.encoder_config.get(str(out_dict.get(col_to_encode)))
+                logger.info('Loaded dictionary', out_dict)
+                f.write(json.dumps(out_dict) + "\n")
+        return
+
+    def run(self):
+        """Preprocessor runner."""
+        if self.user_preprocessor:
+            self.run_user_preprocessor()
+            return
+        if self.encoder_config:
+            self.encoder_config = json.loads(self.encoder_config)
+        if self.template:
+            # all preprocessing will be done based on their provided pattern
+            self.prep_using_template()
+            return
+
+    def run_user_preprocessor(self):
+        """Prerpocessor run using custom template."""
+        try:
+            os.system(
+                f'python {self.user_preprocessor} --input_path {self.input_dataset} \
+                --output_path {self.output_dataset}'
+            )
+        except Exception as e:
+            logger.error('Script failed', e)

```

Comment:
Nit: if you do `logger.exception` it will also include stack trace.

Reply:
Thank you. Added.